engine/runtime-instrument-common/src/main/java/org/enso/interpreter/instrument/RuntimeCache.java

```diff
@@ -177,4 +182,14 @@ public void removeWeight(UUID key) {
   public void clearWeights() {
     weights.clear();
   }
+
+  public void beginQuery(Consumer<UUID> callback) {
+    assert this.observer == null;
+    this.observer = callback;
+  }
+
+  public void finishQuery(Consumer<UUID> callback) {
+    assert this.observer == callback;
+    this.observer = null;
+  }
```

Comment:
I know that currently there is an assert for `null`, but Is it possible that we will want those queries to be reentrant? If so, the `observer` callbacks would need to be stored on a stack (either manully, or through a closure).

Excuse my rusty java, but I'm thinking about something along those lines:

```java
  public void runQuery(Consumer<UUID> callback, Runnable scope) {
    var previousCallback = this.observer;
    this.observer = callback;
    try {
      scope.run()
    }
    finally {
      this.observer = previousCallback;
    }
  }
```

Reply:
Yes, that's better. Thank you: 2cd1c24e38